File: src/encode_scripts/rsa_friend.py
# Name: RSA (Friend)
# Description: 与朋友安全通信 (使用朋友公钥加密，自己私钥解密)
# Avatar: 🤝

from Crypto.PublicKey import RSA
from Crypto.Cipher import PKCS1_OAEP
import base64
import json

# Global config placeholder (injected by JS)
SCRIPT_CONFIG = {}

def get_config_schema():
    """
    Define the configuration UI for this script.
    """
    return json.dumps({
        "params": [
            {
                "name": "friend_public_key",
                "label": "朋友的公钥 (Friend's Public Key) - 用于发送加密消息",
                "type": "textarea",
                "placeholder": "-----BEGIN PUBLIC KEY-----..."
            },
            {
                "name": "my_public_key",
                "label": "我的公钥 (My Public Key) - 发送给朋友",
                "type": "textarea",
                "placeholder": "生成后自动填充，请发送给朋友...",
                "readOnly": True
            },
            {
                "name": "my_private_key",
                "label": "我的私钥 (My Private Key) - 用于解密收到的消息",
                "type": "textarea",
                "placeholder": "生成后自动填充，请勿泄露...",
                "readOnly": True
            }
        ],
        "actions": [
            {
                "name": "generate_keys",
                "label": "生成我的密钥对 (Generate My Keys)",
                "type": "button"
            }
        ]
    })

def generate_keys():
    """
    Generate a new RSA key pair for the user.
    """
    key = RSA.generate(2048)
    private_key = key.export_key().decode('utf-8')
    public_key = key.publickey().export_key().decode('utf-8')
    return json.dumps({
        "my_public_key": public_key,
        "my_private_key": private_key
    })

def encode(text):
    """
    加密函数：使用【朋友的公钥】加密消息，发送给朋友
    """
    if not text:
        return ''
        
    friend_pub_key = SCRIPT_CONFIG.get('friend_public_key', '').strip()
    if not friend_pub_key:
        raise ValueError("请先在设置中配置【朋友的公钥】才能发送加密消息")

    try:
        recipient_key = RSA.import_key(friend_pub_key)
        cipher_rsa = PKCS1_OAEP.new(recipient_key)
        encrypted_msg = cipher_rsa.encrypt(text.encode('utf-8'))
        return base64.b64encode(encrypted_msg).decode('utf-8')
    except ValueError as ve:
         raise ValueError(f"朋友公钥格式错误: {str(ve)}")
    except Exception as e:
         raise ValueError(f"加密错误: {str(e)}")

def decode(text):
    """
    解密函数：使用【我的私钥】解密朋友发来的消息
    """
    if not text:
        return ''
        
    my_priv_key = SCRIPT_CONFIG.get('my_private_key', '').strip()
    if not my_priv_key:
        raise ValueError("请先生成【我的密钥对】才能解密消息")
        
    try:
        private_key = RSA.import_key(my_priv_key)
        cipher_rsa = PKCS1_OAEP.new(private_key)
        
        # Decode base64 first
        try:
            ciphertext = base64.b64decode(text)
        except:
             raise ValueError("输入不是有效的Base64编码")
             
        decrypted_msg = cipher_rsa.decrypt(ciphertext)
        return decrypted_msg.decode('utf-8')
    except ValueError as ve:
         raise ValueError(f"密钥或密文错误: {str(ve)}")
    except Exception as e:
        raise ValueError(f"解密错误: {str(e)} (请确认这是发给您的消息)")

File: src/encode_scripts/rsa_friend.ts
// Name: RSA (Friend)
// Description: 与朋友安全通信 (使用朋友公钥加密，自己私钥解密)
// Avatar: 🤝

function get_config_schema() {
    return {
        params: [
            {
                name: "friend_public_key",
                label: "朋友的公钥 (Friend's Public Key) - 用于发送加密消息",
                type: "textarea",
                placeholder: "-----BEGIN PUBLIC KEY-----..."
            },
            {
                name: "my_public_key",
                label: "我的公钥 (My Public Key) - 发送给朋友",
                type: "textarea",
                placeholder: "生成后自动填充，请发送给朋友...",
                readOnly: true
            },
            {
                name: "my_private_key",
                label: "我的私钥 (My Private Key) - 用于解密收到的消息",
                type: "textarea",
                placeholder: "生成后自动填充，请勿泄露...",
                readOnly: true
            }
        ],
        actions: [
            {
                name: "generate_keys",
                label: "生成我的密钥对 (Generate My Keys)",
                type: "button"
            }
        ]
    };
}

function pemToArrayBuffer(pem) {
    const b64 = pem.replace(/-----[A-Z ]+-----/g, '').replace(/\s/g, '');
    const binary = atob(b64);
    const bytes = new Uint8Array(binary.length);
    for (let i = 0; i < binary.length; i++) {
        bytes[i] = binary.charCodeAt(i);
    }
    return bytes.buffer;
}

function arrayBufferToPem(buffer, type) {
    const bytes = new Uint8Array(buffer);
    let binary = '';
    for (let i = 0; i < bytes.length; i++) {
        binary += String.fromCharCode(bytes[i]);
    }
    const b64 = btoa(binary);
    const lines = b64.match(/.{1,64}/g).join('\n');
    return '-----BEGIN ' + type + '-----\n' + lines + '\n-----END ' + type + '-----';
}

async function generate_keys() {
    const keyPair = await crypto.subtle.generateKey(
        {
            name: 'RSA-OAEP',
            modulusLength: 2048,
            publicExponent: new Uint8Array([1, 0, 1]),
            hash: 'SHA-256'
        },
        true,
        ['encrypt', 'decrypt']
    );
    const publicKey = await crypto.subtle.exportKey('spki', keyPair.publicKey);
    const privateKey = await crypto.subtle.exportKey('pkcs8', keyPair.privateKey);
    return JSON.stringify({
        my_public_key: arrayBufferToPem(publicKey, 'PUBLIC KEY'),
        my_private_key: arrayBufferToPem(privateKey, 'PRIVATE KEY')
    });
}

async function encode(text) {
    if (!text) return '';

    const friendPubKey = (SCRIPT_CONFIG.friend_public_key || '').trim();
    if (!friendPubKey) {
        throw new Error("请先在设置中配置【朋友的公钥】才能发送加密消息");
    }

    try {
        const keyData = pemToArrayBuffer(friendPubKey);
        const key = await crypto.subtle.importKey(
            'spki', keyData,
            { name: 'RSA-OAEP', hash: 'SHA-256' },
            false, ['encrypt']
        );
        const encrypted = await crypto.subtle.encrypt(
            { name: 'RSA-OAEP' },
            key,
            new TextEncoder().encode(text)
        );
        const bytes = new Uint8Array(encrypted);
        let binary = '';
        for (let i = 0; i < bytes.length; i++) {
            binary += String.fromCharCode(bytes[i]);
        }
        return btoa(binary);
    } catch (e) {
        const msg = e.message || String(e);
        if (msg.indexOf("请先") !== -1) throw e;
        throw new Error("加密错误: " + msg);
    }
}

async function decode(text) {
    if (!text) return '';

    const myPrivKey = (SCRIPT_CONFIG.my_private_key || '').trim();
    if (!myPrivKey) {
        throw new Error("请先生成【我的密钥对】才能解密消息");
    }

    try {
        let binary;
        try {
            binary = atob(text);
        } catch (e) {
            throw new Error("输入不是有效的Base64编码");
        }

        const ciphertext = new Uint8Array(binary.length);
        for (let i = 0; i < binary.length; i++) {
            ciphertext[i] = binary.charCodeAt(i);
        }

        const keyData = pemToArrayBuffer(myPrivKey);
        const key = await crypto.subtle.importKey(
            'pkcs8', keyData,
            { name: 'RSA-OAEP', hash: 'SHA-256' },
            false, ['decrypt']
        );
        const decrypted = await crypto.subtle.decrypt(
            { name: 'RSA-OAEP' },
            key,
            ciphertext
        );
        return new TextDecoder().decode(decrypted);
    } catch (e) {
        const msg = e.message || String(e);
        if (msg.indexOf("请先") !== -1 || msg.indexOf("请先生成") !== -1 || msg.indexOf("输入不是") !== -1) throw e;
        throw new Error("解密错误: " + msg + " (请确认这是发给您的消息)");
    }
}
